packages/core/src/RenderingEngine/WSIViewport.ts

```diff
@@ -0,0 +1,533 @@
+    if (canvas.width !== clientWidth || canvas.height !== clientHeight) {
+      canvas.width = clientWidth;
+      canvas.height = clientHeight;
+    }
+
+    this.refreshRenderValues();
+  };
+
+  /**
+   * Converts a VideoViewport canvas coordinate to a video coordinate.
+   *
+   * @param canvasPos - to convert to world
+   * @returns World position
+   */
+  public canvasToWorld = (canvasPos: Point2): Point3 => {
+    const pan: Point2 = this.videoCamera.panWorld; // In world coordinates
+    const worldToCanvasRatio: number = this.getWorldToCanvasRatio();
+
+    const panOffsetCanvas: Point2 = [
+      pan[0] * worldToCanvasRatio,
+      pan[1] * worldToCanvasRatio,
+    ];
+
+    const subCanvasPos: Point2 = [
+      canvasPos[0] - panOffsetCanvas[0],
+      canvasPos[1] - panOffsetCanvas[1],
+    ];
+
+    const worldPos: Point3 = [
+      subCanvasPos[0] / worldToCanvasRatio,
+      subCanvasPos[1] / worldToCanvasRatio,
+      0,
+    ];
+
+    return worldPos;
+  };
+
+  public async setWSI(imageIds: string[], client) {
+    console.log('Setting image ids', imageIds.length);
+    this.microscopyElement.style.background = 'red';
+    this.microscopyElement.innerText = 'Loading';
+    this.imageIds = imageIds;
+    const DicomMicroscopyViewer = await import('dicom-microscopy-viewer');
+    console.log('Loaded DICOMMicroscopyViewer', DicomMicroscopyViewer);
+    this.frameOfReferenceUID = null;
+    const metadata = this.imageIds.map((imageId) => {
+      const imageMetadata = client.getDICOMwebMetadata(imageId);
+
+      Object.defineProperty(imageMetadata, 'isMultiframe', {
+        value: imageMetadata.isMultiframe,
+        enumerable: false,
+      });
+      Object.defineProperty(imageMetadata, 'frameNumber', {
+        value: undefined,
+        enumerable: false,
+      });
+      const imageType = imageMetadata['00080008']?.Value;
+      if (imageType?.length === 1) {
+        imageMetadata['00080008'].Value = imageType[0].split('\\');
+      }
+      const frameOfReference = imageMetadata['00200052']?.Value?.[0];
+      if (!this.frameOfReferenceUID) {
+        this.frameOfReferenceUID = frameOfReference;
+      } else if (frameOfReference !== this.frameOfReferenceUID) {
+        imageMetadata['00200052'].Value = [this.frameOfReferenceUID];
+      }
+
+      return imageMetadata;
+    });
+    console.log('Loaded metadata for', metadata.length, 'images');
+    const volumeImages = [];
+    metadata.forEach((m) => {
+      const image =
+        new DicomMicroscopyViewer.metadata.VLWholeSlideMicroscopyImage({
+          metadata: m,
+        });
+      const imageFlavor = image.ImageType[2];
+      if (imageFlavor === 'VOLUME' || imageFlavor === 'THUMBNAIL') {
+        volumeImages.push(image);
+      } else {
+        console.log('Image type not volume', image.ImageType);
+      }
+    });
+    // Construct viewer instance
+    const viewer = new DicomMicroscopyViewer.viewer.VolumeImageViewer({
```

Comment:
didn't know the api for DMV has VolumeImageViewer, what is volume for WSI? the tiles?

Reply:
Will have to look it up - it might be the fact that it actually can be a volume, with various focal depths.

Reply:
I'm not adding focal depth support right now, although it is something we might want to consider.